test/e2e/aks_tags.go

```diff
@@ -0,0 +1,172 @@
+	subscriptionID := settings.GetSubscriptionID()
+	auth, err := settings.GetAuthorizer()
+	Expect(err).NotTo(HaveOccurred())
+
+	managedclustersClient := containerservice.NewManagedClustersClient(subscriptionID)
+	managedclustersClient.Authorizer = auth
+
+	agentpoolsClient := containerservice.NewAgentPoolsClient(subscriptionID)
+	agentpoolsClient.Authorizer = auth
+
+	mgmtClient := bootstrapClusterProxy.GetClient()
+	Expect(mgmtClient).NotTo(BeNil())
+
+	infraControlPlane := &infrav1exp.AzureManagedControlPlane{}
+	err = mgmtClient.Get(ctx, client.ObjectKey{
+		Namespace: input.Cluster.Spec.ControlPlaneRef.Namespace,
+		Name:      input.Cluster.Spec.ControlPlaneRef.Name,
+	}, infraControlPlane)
+	Expect(err).NotTo(HaveOccurred())
+
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer GinkgoRecover()
+		defer wg.Done()
+
+		By("Deleting all tags for control plane")
+		initialTags := infraControlPlane.Spec.AdditionalTags
+		Expect(initialTags).NotTo(BeNil())
+		infraControlPlane.Spec.AdditionalTags = nil
+		Expect(mgmtClient.Update(ctx, infraControlPlane)).To(Succeed())
+
+		Eventually(func(g Gomega) {
+			aksCluster, err := managedclustersClient.Get(ctx, infraControlPlane.Spec.ResourceGroupName, infraControlPlane.Name)
+			g.Expect(err).NotTo(HaveOccurred())
+			g.Expect(aksCluster.Tags).To(BeNil())
+		}, input.WaitForUpdate...).Should(Succeed())
+
+		By("Creating tags for control plane")
+		expectedTags := infrav1.Tags{
+			"test":    "tag",
+			"another": "value",
+		}
+		Expect(mgmtClient.Get(ctx, client.ObjectKeyFromObject(infraControlPlane), infraControlPlane)).To(Succeed())
+		infraControlPlane.Spec.AdditionalTags = expectedTags
+		Expect(mgmtClient.Update(ctx, infraControlPlane)).To(Succeed())
+
+		checkTags := func(g Gomega) {
+			aksCluster, err := managedclustersClient.Get(ctx, infraControlPlane.Spec.ResourceGroupName, infraControlPlane.Name)
+			g.Expect(err).NotTo(HaveOccurred())
+			actualTags := converters.MapToTags(aksCluster.Tags)
+			g.Expect(actualTags).To(Equal(expectedTags))
+		}
+
+		Eventually(checkTags, input.WaitForUpdate...).Should(Succeed())
+
+		By("Updating tags for control plane")
+		expectedTags = initialTags
+		Expect(mgmtClient.Get(ctx, client.ObjectKeyFromObject(infraControlPlane), infraControlPlane)).To(Succeed())
+		infraControlPlane.Spec.AdditionalTags = expectedTags
+		Expect(mgmtClient.Update(ctx, infraControlPlane)).To(Succeed())
+
+		Eventually(checkTags, input.WaitForUpdate...).Should(Succeed())
+	}()
+
+	for _, mp := range input.MachinePools {
+		wg.Add(1)
+		go func(mp *expv1.MachinePool) {
+			defer GinkgoRecover()
+			defer wg.Done()
+
+			Byf("Creating tags for machine pool %s", mp.Name)
+			ammp := &infrav1exp.AzureManagedMachinePool{}
+			Expect(mgmtClient.Get(ctx, types.NamespacedName{
+				Namespace: mp.Spec.Template.Spec.InfrastructureRef.Namespace,
+				Name:      mp.Spec.Template.Spec.InfrastructureRef.Name,
+			}, ammp)).To(Succeed())
+			Expect(ammp.Spec.AdditionalTags).To(BeNil())
```

Comment:
Is this base case durable enough? Should we clear out the tags first of all like we're doing for control plane so that this scenario will work w/ a cluster spec that ships w/ AdditionalTags in an AzureManagedMachinePool out of the gate?

Reply:
Yeah that seems like a good idea. I'll add that here on the front end in addition to the "delete all" later since we know for sure at that point that there are tags to delete. I might tweak the corresponding test against AzureManagedControlPlane too to be more symmetrical with this one too.

Reply:
Cool. Super excited to land these, this is awesome.

Reply:
Just pushed this change. Now the tests against AzureManagedControlPlane and AzureManagedMachinePool are doing the exact same sequence of operations, roughly:
1. Delete all the existing tags
2. Create some new tags
3. Update the tags
4. Restore the initial tags

Whether or not tags were initially present, this should cover the main scenarios. The only slightly unintuitive nuance is that when no tags are initially defined, the "delete all tags" step is basically a no-op and the "restore initial tags" step is really testing the delete.

Reply:
sgtm

I think the coverage of "no op delete all tags" is still an interesting scenario fwiw.